### ISSUE TITLE
Blind accessability keyboard navigation 

### ISSUE THREAD

**Issue by Dirtypawz** [WeakAuras/WeakAuras2#4585]
Hello

I spoke with WA over twitter and gave a brief idea of the request and they also asked me to post it here.
Im totally blind and I’ve raided in wow since wrath, Google Ben shaw blind wow raider so you know it’s not a troll.

I would like to request keyboard navigation of WA is looked at to be added so myself and others can independently use WA to its full capacity for making auras with sound alerts.
TTS was added to wow as you know but another addon expanded TTS capability to read anything in the UI including other addons, it’s called keyboard UI. Currently I’m able to open WA and have anything my cursor goes over read out, unfortunately relying on the cursor read out to make auras whilst totally blind doesn’t work out well, WA adding keyboard navigation would make it possible.
Keyboard UI enables navigation of the spellbook, bars, bags and the ability to place spells on the bars. I think keyboard UI would be a good reference for the navigation keys and simple but very effective way we can acces UI elements through the keyboard.

wa is critical to being able to raid and do M+ blind, adding keyboard navigation would allow us to utilise WA further as we wouldn’t be set by the bounds of relying on sighted help to do things and also some blind players just don’t have access to sighted help so it would open a whole new level of play to them.

thank you

**Comment by InfusOnWoW** [comment]
How is keyboard ui enabled in the wow client?

**Comment by Dirtypawz** [comment]
Hi So keyboard UI is a seporate addon available on curse.It should just be install and work.I do not think keyboard UI shows a visual thing when navigating the UI and it may require enabling wows TTS.Hope this helps Sent from my iPhoneOn 2 Sep 2023, at 18:22, InfusOnWoW ***@***.***> wrote:﻿
How is keyboard ui enabled in the wow client?

**Comment by Dirtypawz** [comment]
https://www.curseforge.com/wow/addons/keyboardui/downloadSent from my iPhoneOn 2 Sep 2023, at 18:22, InfusOnWoW ***@***.***> wrote:﻿
How is keyboard ui enabled in the wow client?

**Comment by InfusOnWoW** [comment]
Is there any documentation on that addon? What kind of work would we need to do?

 I do not want to read the source of ARR addons, because reading that is against the wishes of the author.

**Comment by Dirtypawz** [comment]
So the keyboard navigation of WA wouldn’t come from keyboard UI, it would come from yourselves adding it as a function. There are several other addons that have added its own keyboard navigation within its addon UI, notably blind slash and command Palet.The only relavence of keyboard UI in this is that it expands wows TTS to read everything in the UI including addons when I hover over things with the cursor, I don’t know if this would change when keyboard navigation of WA is added.I predominantly brought up keyboard UI as it is a example of how keyboard navigation is used by us if you needed a point of reference and I was asked by the WA person on twitter if I had a example of keyboard navigation in a addon.Essentially keyboard navigation would have to be added by you for use within WA.Sent from my iPhoneOn 2 Sep 2023, at 20:00, InfusOnWoW ***@***.***> wrote:﻿
Is there any documentation on that addon? What kind of work would we need to do?
I do not want to read the source of ARR addons, because reading that is against the wishes of the author.

**Comment by InfusOnWoW** [comment]
So I looked at how keyboard ui does it. And the addons purpose is basically to add some minimal accesibility to the base ui.
It does two main things:
* Mouse over tts for in game objects and ui elements
* Modify blizzard uis to add keyboard navigation

Now, maybe it was because I was testing on retail instead of classic or woltk, but the keyboard navigation felt pretty yanky.

Anyway the shortcuts were:
* Ctrl+Up/Down to navigate
* Ctrl+Left/Right/Enter to activate controls
* Ctrl+Space to read out the current element that has focus
* Alt+Space to read out extra actions, dialogs the bottom buttons are always mapped to extra actions
* Alt+# to activate an extra action

So to support that we would need to:
Add some kind of "focus" concept to widgets, which works for all widgets not just editboxes
Add some kind of "focus-chain" concept.
Add the keyboard shortcuts for activating controls, reading out meta information, and focussing the next/prev widget

We use AceGui+AceOptions though, which adds further complications. As we would need to add the support for that to AceGui, then make AceOptions expose an api that abstracts over that.

I'm afraid that's a rather lot of code that needs to be written, and I think it's unlikely that will ever happen.

**Comment by Stanzilla** [comment]
@Nevcairiel is that something you guys would be open to adding to Ace3 in general?

**Comment by Dirtypawz** [comment]
HiThank you for getting back.I understand, i thought I would ask as I would never know if it was possible if I didn’t.Thank you so much for taking the time to look.I love WA as it enables me to do so much more. All the best for the future.Sent from my iPhoneOn 6 Sep 2023, at 01:32, Benjamin Staneck ***@***.***> wrote:﻿
@Nevcairiel is that something you guys would be open to adding to Ace3 in general?

**Comment by Stanzilla** [comment]
> HiThank you for getting back.I understand, i thought I would ask as I would never know if it was possible if I didn’t.Thank you so much for taking the time to look.I love WA as it enables me to do so much more. All the best for the future.Sent from my iPhoneOn 6 Sep 2023, at 01:32, Benjamin Staneck ***@***.***> wrote:﻿ @Nevcairiel is that something you guys would be open to adding to Ace3 in general? —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

Thank you for bringing it up! Accessibility is an area where we all need to improve.

**Comment by InfusOnWoW** [comment]
I did spent an afternoon looking one more time at this by coding and getting a better feel for how much work this would take.
I actually made more progress than I expected. But I also found that there are lots of places that would be very difficult to handle.

I got tabbing to mostly work, although most widgets aren't read-out correctly.

I did upload what I have to https://github.com/WeakAuras/WeakAuras2/pull/4607 and created a package from that here:
https://github.com/WeakAuras/WeakAuras2/actions/runs/6210192265?pr=4607

With that experiment I can say with confidence that this is unfortunately far too much effort to realistically get done.

**Comment by InfusOnWoW** [comment]
Ah well, I though I could polish it a bit more so that it's in a demo/proof of concept state, where maybe someone else is interested in doing the actual hard work.

That code is at: https://github.com/InfusOnWoW/WeakAuras2/tree/tts-experiment
And a demo of what it does is at: https://www.youtube.com/watch?v=sq6A1PJlX3w

Obviously a video isn't exactly the best medium for demonstrating accessibility for blind users, but eh that's what I could easily produce.

I should stress, this is a demo. Even though it appears that lots of things work, that's only because I navigate compentely towards the things that do work. Tabbing mostly works, though the implementation needs work. Activating controls only works for checkboxes. Some widgets would be easy to add, others pretty hard.

In any case @Stanzilla maybe we can tweet and ask whether anyone wants to tackle this?

**Comment by Dirtypawz** [comment]
Hi allThank you very much for looking further into this. I get that it’s more work on top of what is probably already a mountain load with making WA as good as it is.Best regards, Ben.Sent from my iPhoneOn 17 Sep 2023, at 12:57, InfusOnWoW ***@***.***> wrote:﻿
Ah well, I though I could polish it a bit more so that it's in a demo/proof of concept state, where maybe someone else is interested in doing the actual hard work.
That code is at: https://github.com/InfusOnWoW/WeakAuras2/tree/tts-experiment
And a demo of what it does is at: https://www.youtube.com/watch?v=sq6A1PJlX3w
Obviously a video isn't exactly the best medium for demonstrating accessibility for blind users, but eh that's what I could easily produce.
I should stress, this is a demo. Even though it appears that lots of things work, that's only because I navigate compentely towards the things that do work. Tabbing mostly works, though the implementation needs work. Activating controls only works for checkboxes. Some widgets would be easy to add, others pretty hard.
In any case @Stanzilla maybe we can tweet and ask whether anyone wants to tackle this?

**Comment by Dirtypawz** [comment]
Again to all at WA, thank you for taking the time to look, code and upload a video on this, it is greatly appreciated.Sent from my iPhoneOn 17 Sep 2023, at 12:57, InfusOnWoW ***@***.***> wrote:﻿
Ah well, I though I could polish it a bit more so that it's in a demo/proof of concept state, where maybe someone else is interested in doing the actual hard work.
That code is at: https://github.com/InfusOnWoW/WeakAuras2/tree/tts-experiment
And a demo of what it does is at: https://www.youtube.com/watch?v=sq6A1PJlX3w
Obviously a video isn't exactly the best medium for demonstrating accessibility for blind users, but eh that's what I could easily produce.
I should stress, this is a demo. Even though it appears that lots of things work, that's only because I navigate compentely towards the things that do work. Tabbing mostly works, though the implementation needs work. Activating controls only works for checkboxes. Some widgets would be easy to add, others pretty hard.
In any case @Stanzilla maybe we can tweet and ask whether anyone wants to tackle this?
